druid-shell/src/platform/x11/keycodes.rs

```diff
@@ -0,0 +1,335 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! X11 keycode handling.
+
+use gdk::enums::key::*;
+
+use crate::keyboard::StrOrChar;
+use crate::keycodes::KeyCode;
+
+pub type RawKeyCode = u32;
+
+impl From<RawKeyCode> for KeyCode {
+    #[allow(clippy::just_underscores_and_digits, non_upper_case_globals)]
+    fn from(raw: u32) -> KeyCode {
+        match raw {
+            9 => KeyCode::Escape,
```

Comment:
for these kinds of magic numbers I find it helpful to point to some documentation source that supports the mapping used.

Reply:
Gonna have to defer to @perlindgren here